pkg/scale/decode_test.go

```diff
@@ -0,0 +1,169 @@
+				t.Errorf("decodeState.unmarshal() = %v, want %v", dst, tt.in)
+			}
+		})
+	}
+}
+
+func Test_decodeState_decodeBytes(t *testing.T) {
+	for _, tt := range stringTests {
+		t.Run(tt.name, func(t *testing.T) {
+			dst := reflect.New(reflect.TypeOf(tt.in)).Elem().Interface()
+			if err := Unmarshal(tt.want, &dst); (err != nil) != tt.wantErr {
+				t.Errorf("decodeState.unmarshal() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(dst, tt.in) {
+				t.Errorf("decodeState.unmarshal() = %v, want %v", dst, tt.in)
+			}
+		})
+	}
+}
+
+func Test_decodeState_decodeBool(t *testing.T) {
+	for _, tt := range boolTests {
+		t.Run(tt.name, func(t *testing.T) {
+			var dst bool
+			if err := Unmarshal(tt.want, &dst); (err != nil) != tt.wantErr {
+				t.Errorf("decodeState.unmarshal() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !reflect.DeepEqual(dst, tt.in) {
+				t.Errorf("decodeState.unmarshal() = %v, want %v", dst, tt.in)
+			}
+		})
+	}
+}
+
+func Test_decodeState_decodeStruct(t *testing.T) {
+	for _, tt := range structTests {
+		t.Run(tt.name, func(t *testing.T) {
+			dst := reflect.New(reflect.TypeOf(tt.in)).Elem().Interface()
+			if err := Unmarshal(tt.want, &dst); (err != nil) != tt.wantErr {
+				t.Errorf("decodeState.unmarshal() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			var diff string
+			if tt.out != nil {
+				diff = cmp.Diff(dst, tt.out, cmpopts.IgnoreUnexported(tt.in))
+			} else {
+				diff = cmp.Diff(dst, tt.in, cmpopts.IgnoreUnexported(big.Int{}, tt.in, VDTValue2{}, MyStructWithIgnore{}))
+			}
+			if diff != "" {
+				t.Errorf("decodeState.unmarshal() = %s", diff)
+			}
+		})
+	}
+}
+func Test_decodeState_decodeArray(t *testing.T) {
+	for _, tt := range arrayTests {
+		t.Run(tt.name, func(t *testing.T) {
+			dst := reflect.New(reflect.TypeOf(tt.in)).Elem().Interface()
+			if err := Unmarshal(tt.want, &dst); (err != nil) != tt.wantErr {
+				t.Errorf("decodeState.unmarshal() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !reflect.DeepEqual(dst, tt.in) {
+				t.Errorf("decodeState.unmarshal() = %v, want %v", dst, tt.in)
+			}
+		})
+	}
+}
+
+func Test_decodeState_decodeSlice(t *testing.T) {
+	for _, tt := range sliceTests {
+		t.Run(tt.name, func(t *testing.T) {
+			dst := reflect.New(reflect.TypeOf(tt.in)).Elem().Interface()
+			if err := Unmarshal(tt.want, &dst); (err != nil) != tt.wantErr {
+				t.Errorf("decodeState.unmarshal() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if !reflect.DeepEqual(dst, tt.in) {
+				t.Errorf("decodeState.unmarshal() = %v, want %v", dst, tt.in)
+			}
+		})
+	}
+}
+
+func Test_unmarshal_optionality(t *testing.T) {
+	var ptrTests tests
+	for _, t := range allTests {
+		ptrTest := test{
+			name:    t.name,
+			in:      t.in,
+			wantErr: t.wantErr,
+			want:    t.want,
+			out:     t.out,
+		}
+		switch t.in {
+		case nil:
+			// this doesn't actually happen since none of the tests have nil value for tt.in
```

Comment:
we should probably test the nil case for this

Reply:
added new nil test cases for all test cases.  Maybe overkill, since it's testing the same type, but I'd rather do that than cherry pick tests from each subset of tests.